bin/propolis-server/src/lib/spec/api_spec_v0.rs

```diff
@@ -0,0 +1,259 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Conversions from version-0 instance specs in the [`propolis_api_types`]
+//! crate to the internal [`super::Spec`] representation.
+
+use propolis_api_types::instance_spec::{
+    components::devices::{SerialPort as SerialPortDesc, SerialPortNumber},
+    v0::{InstanceSpecV0, NetworkBackendV0, NetworkDeviceV0, StorageDeviceV0},
+};
+use thiserror::Error;
+
+#[cfg(feature = "falcon")]
+use propolis_api_types::instance_spec::components::devices::SoftNpuPort as SoftNpuPortSpec;
+
+#[cfg(feature = "falcon")]
+use crate::spec::SoftNpuPort;
+
+use super::{
+    builder::{SpecBuilder, SpecBuilderError},
+    Disk, Nic, QemuPvpanic, SerialPortUser, Spec,
+};
+
+#[derive(Debug, Error)]
+pub(crate) enum ApiSpecParseError {
+    #[error(transparent)]
+    BuilderError(#[from] SpecBuilderError),
+
+    #[error("backend {backend} not found for device {device}")]
+    BackendNotFound { backend: String, device: String },
+
+    #[error("backend {0} not used by any device")]
+    BackendNotUsed(String),
+
+    #[error("network backend for guest NIC {0} is not a viona backend")]
+    GuestNicInvalidBackend(String),
+
+    #[cfg(feature = "falcon")]
+    #[error("network backend for device {0} is not a DLPI backend")]
+    NotDlpiBackend(String),
+}
+
+impl From<Spec> for InstanceSpecV0 {
+    fn from(val: Spec) -> Self {
+        let mut spec = InstanceSpecV0::default();
+
+        spec.devices.board = val.board;
+        for (disk_name, disk) in val.disks {
+            let _old = spec
+                .devices
+                .storage_devices
+                .insert(disk_name, disk.device_spec.into());
+
+            assert!(_old.is_none());
+
+            let _old = spec
+                .backends
+                .storage_backends
+                .insert(disk.backend_name, disk.backend_spec.into());
+
+            assert!(_old.is_none());
+        }
+
+        for (nic_name, nic) in val.nics {
+            let _old = spec
+                .devices
+                .network_devices
+                .insert(nic_name, NetworkDeviceV0::VirtioNic(nic.device_spec));
+
+            assert!(_old.is_none());
+
+            let _old = spec.backends.network_backends.insert(
+                nic.backend_name,
+                NetworkBackendV0::Virtio(nic.backend_spec),
+            );
+
+            assert!(_old.is_none());
```

Comment:
tiny style nit: i don't think the `_` to suppress unused-variable warnings for `_old` is necessary here, since we use a regular `assert!` rather than a `debug_assert!` --- the value of `_old` actually *is* always used.
```suggestion
            let old = spec
                .devices
                .storage_devices
                .insert(disk_name, disk.device_spec.into());

            assert!(old.is_none());

            let old = spec
                .backends
                .storage_backends
                .insert(disk.backend_name, disk.backend_spec.into());

            assert!(old.is_none());
        }

        for (nic_name, nic) in val.nics {
            let old = spec
                .devices
                .network_devices
                .insert(nic_name, NetworkDeviceV0::VirtioNic(nic.device_spec));

            assert!(old.is_none());

            let old = spec.backends.network_backends.insert(
                nic.backend_name,
                NetworkBackendV0::Virtio(nic.backend_spec),
            );

            assert!(old.is_none());
```

Reply:
Also, I was going to ask whether it was correct to assert here, or if a duplicate name should be a user-facing error...but, looking at the `Spec` type, these are all stored in `HashMap`s, so there should never be duplicate names in the `Spec`. Perhaps it's worth documenting that, either with a comment or (preferably) in the assertion's panic message? It's probably obvious to you as the author of this code, but as a reader, I had to go check the `Spec` type...

Not a big deal though.

Reply:
Also, maybe we should write `assert_eq!(old, None)` so that the panic message includes the value of `old`, and/or include the duplicated name? In practice, it doesn't really matter, because this shouldn't ever happen...

Reply:
Took me a couple of tries to get this where I wanted it, but this should be fixed in 02be2a2.